src/main/java/com/fasterxml/jackson/databind/util/TokenBuffer.java

```diff
@@ -1220,18 +1220,32 @@ private void _copyBufferValue(JsonParser p, JsonToken t) throws IOException
             case BIG_INTEGER:
                 writeNumber(p.getBigIntegerValue());
                 break;
-            default:
+            case LONG:
                 writeNumber(p.getLongValue());
+                break;
+            default:
+                throw new RuntimeException(
+                        "Internal error: unexpected VALUE_NUMBER_INT type: " + p.getNumberType());
             }
             break;
         case VALUE_NUMBER_FLOAT:
             if (_forceBigDecimal) {
                 writeNumber(p.getDecimalValue());
             } else {
-                // 09-Jul-2020, tatu: Used to just copy using most optimal method, but
-                //  issues like [databind#2644] force to use exact, not optimal type
-                final Number n = p.getNumberValueExact();
-                _appendValue(JsonToken.VALUE_NUMBER_FLOAT, n);
```

Comment:
This is where NumberType.DOUBLE input is transformed into NumberType.BIG_DECIMAL. My change breaks  the two referenced test cases in JDKNumberDeserTest.java where I've had to enable USE_BIG_DECIMAL_FOR_FLOATS.

Reply:
Ahhhhhh. This is interesting... this exposes an unplanned discrepancy in logic at parser level and requires some more discussion actually.

Problem is as follows:

1. JSON floating-point numbers do not really have anything to tell "minimal" type (unlike integers where we can figure it out)
2. Answers for "true" floating-point type should be consistent -- `getNumberType()` and `getNumberValueExact()` should operate similarly -- but latter takes "more accurate" position, whereas former defaults to `NumberType.DOUBLE` (for historical reasons)

Given this, I am not really comfortable with change here. This is not a new dilemma, #1770 seems related, but in general what would really be needed, ideally, would be for JSON (and other non-strictly-typed formats) to defer actual decoding until value is needed. But I have no idea how that could actually be implemented.

But I can see why double<->BigDecimal case can be fluid so maybe earlier patch makes sense.

Reply:
Agreed. I think the trouble writing the deferred implementation is that it requires every dataformat to handle it separately. Happy closing this out in favor of the other approach.

I can remove the long changes from the other PR given they don't appear to be impacted.